include/picongpu/plugins/transitionRadiation/Calculator.hpp

```diff
@@ -0,0 +1,238 @@
+            float_X const b = particle.getU( ) * parMomSinTheta * parMomCosPhi;
+            float_X const c = parSqrtOnePlusUSquared * detectorSinTheta;
+
+            // Denominator
+            float_X const x = parSqrtOnePlusUSquared - 
+                particle.getU( ) * parMomSinTheta * parMomCosPhi * detectorSinTheta;
+            float_X const y = particle.getU( ) * parMomCosTheta * detectorCosTheta;
+
+            float_X denominator = x * x - y * y; 
+
+            // Preventing division by 0
+            if( math::abs( denominator ) < DIV_BY_ZERO_MINIMUM )
+            {
+                if( denominator < 0.0 )
+                    denominator = -DIV_BY_ZERO_MINIMUM;
+                else
+                    denominator = DIV_BY_ZERO_MINIMUM;
+            }
+
+            return a * ( b - c ) * ( 1.0 / denominator );
+        }
+
+        /** Exponent of form factor
+         * 
+         * Calculates the exponent of the formfactor divided by \omega
+         * It represents the phase of a single electron in the bunch, but it is mostly
+         * calculated for performance reasons.
+         * \f[ F_exp = - i z ( 1 / v - \sin{\theta} \sin{\psi} \cos{\phi_P - \phi_D} / c ) / \cos{\phi}
+         *          - i \sin{\theta} \rho \cos{\phi_P - \phi_D} \f]
+         * 
+         */
+        HDINLINE
+        complex_X 
+        calcFormFactorExponent( ) const
+        {
+            // If case for longitudinal moving particles... leads to 0 later in the kernel
+            if ( math::abs( parMomCosTheta ) <= DIV_BY_ZERO_MINIMUM )
+                return complex_X( -1.0, 0.0 );
+                
+            float_X const a = detectorSinTheta * parMomSinTheta * math::cos( parMomPhi - detectorPhi );
+            float_X const b = - ( particle.getPosPara( ) ) * ( 1 / particle.getVel( ) - a / SPEED_OF_LIGHT) / ( parMomCosTheta );
+            float_X const c = - detectorSinTheta * particle.getPosPerp( ) * math::cos( particle.getPosPhi( ) - detectorPhi );
+
+            complex_X const fpara = complex_X( 0.0, b );
+            complex_X const fperp = complex_X( 0.0, c );
+            return fpara + fperp;
+            
+        }
+    }; // class Calculator
+
+    /** Formfactor
+     * 
+     * Calculates of the electron bunch with the exponent calculated by the 
+     * Calculator class. 
+     * 
+     * @f[F = \exp{ F_{exp} * \omega }@f]
+     * 
+     * @param omega observed frequency
+     * @param exponent exponent of exponential function
+     */
+    HDINLINE
+    complex_X 
+    calcFormFactor(
+        float_X const omega, 
+        complex_X const exponent
+    )
+    {
+        // If case for longitudinal moving particles
+        if ( exponent.get_real() == -1.0 )
+            return complex_X( 0.0, 0.0 );
+        else
+            return complex_X( 
+                math::exp( 
+                    //precisionCast< float_64 >( exponent * omega )
+                    exponent * omega
```

Comment:
Here you cloud avoid branching too:
```c++
const bool long_mov_particle = exponent.get_real() == -1.0;
int(long_mov_particle) * complex_X( 0.0, 0.0 ) + int(!long_mov_particle) * complex_X( ... ;
```

Reply:
@psychocoderHPC Could you please comment whether such optimizations are useful here?